### PR TITLE
Rework handling of floating window geometry

### PR DESCRIPTION
**Pull request by Stebalien**
This fixes several bugs with respect to the placement/size of floating windows.

1. EXWM is now compatible with spacious-padding-mode.
2. Initial placement/size of floating frames is fixed. It used to be slightly off causing windows to repeatedly change size as they were re-opened over and over. Many programs remember the sizes of their windows to put them back where the user left them. Slightly shifting window locations/sizes on open will cause them to "walk". 3. All manual calculations (header line, mode line, etc.) with respect to window sizes have been removed in favor of letting Emacs tell us what he correct numbers are.

Additionally, this commit contains a fair amount of code cleanup (preferring cond over nested conditions, replacing * and ** variable names in favor of descriptive names, etc.).

### CODE REVIEW

File: exwm-floating.el [hunk 1]
```diff
@@ -146,129 +143,56 @@ If TILED-P is non-nil, set actions for tiled window."
...
-         ;; Create new frame
-         (frame (with-current-buffer
-                    (or (get-buffer "*scratch*")
-                        (progn
-                          (set-buffer-major-mode
-                           (get-buffer-create "*scratch*"))
-                          (get-buffer "*scratch*")))
-                  (make-frame
+  ;; Hide the non-floating X window first.
+  (when-let* ((window (get-buffer-window (exwm--id->buffer id) 'visible)))
```
[the remark this answers is not in the document]
Comment (medranocalvo) [reply exwm-floating.el]: > X window buffers can only appear on one frame, IIRC.

This is a restriction I'd like to lift, to the extent possible.  Nonetheless, I've got no code so we can deal with that when/if it happens.

> I changed this to avoid assuming that the X window is on the current workspace. TBH, I think it might be even safer to pass `t`.

* When the user presses the key combination, the X window of the `exwm-mode` buffer in the selected window on the selected frame should float.  Given that the `exwm-mode` buffer can only be displayed in a single window globally, this should work.
* When a new X window is managed, it can be set to float via the `exwm-configuration`.  Can one configure an X window to appear at a particular workspace?  In that case we should ensure that the window floats over that particular frame and not over currently selected frame.

> Alternatively, any opinions on calling `replace-buffer-in-windows` instead? That'll kill dedicated windows before making X window float.

From the outset I'd say we should respect window dedication.  As long as `exwm-mode` buffers can only be displayed in one window or making them floating extracts them from the window they are being displayed into the new floating frame we should prevent floating when the window the `exwm-mode` buffer is displayed into is dedicated to it.  This is not a feature I use, so my view might be myopic.

**Comment by Stebalien** [comment]
> Generally I am okay if you just merge your changes directly if you are
reasonably confident that they are correct.

Yeah, I tend to do that for simple changes but this change had a lot of platform edge-cases so I wanted to give people a chance to test and give feedback. I'll merge it now as it seems to be good to go.

> Regarding the GH review interface

The GH review interface is awful and slow. It's better if you use a chromium based browser (and make sure to keep VM optimizations on) but it's still slow and power hungry.

I used the `pr-review` package for a while but but it was never stable enough in my experience.

**Comment by Stebalien** [comment]
> Yes, I use Qutebrowser right now since it integrates nicely with EXWM and it is based on Chromium. Something must have changed which made the site horribly inefficient.

Ah, well, you could try profiling with the dev tools. It could be misbehaving browser feature detection (Google was famous for doing this to Firefox), a content blocker and/or missing feature that's causing the website to constantly "retry" an operation (I've hit this before).
